os_info/src/linux/file_release.rs

```diff
@@ -31,100 +31,158 @@ fn retrieve(distributions: &[ReleaseInfo]) -> Option<Info> {
             continue;
         }
 
-        let os_type = Matcher::KeyValue { key: "NAME" }
-            .find(&file_content)
-            .and_then(|name| get_type(&name))
-            .unwrap_or(release_info.os_type);
+        let os_type = (release_info.os_type)(&file_content);
 
-        let version = release_info
-            .version_matcher
-            .find(&file_content)
-            .map(Version::from_string)
-            .unwrap_or_else(|| Version::Unknown);
+        // If os_type is indeterminate, try the next release_info
+        if let None = os_type {
+            continue;
+        }
+
+        let version = (release_info.version)(&file_content);
 
         return Some(Info {
-            os_type,
-            version,
+            os_type: os_type.unwrap(),
+            version: version.unwrap_or_else(|| Version::Unknown),
             bitness: Bitness::Unknown,
             ..Default::default()
         });
     }
 
+    // Failed to determine os info
     None
 }
 
-fn get_type(name: &str) -> Option<Type> {
-    let name = name.to_lowercase();
-
-    // RHEL sometimes has a suffix, like Server or Workstation
-    if name.starts_with("red hat enterprise linux") {
-        return Some(Type::RedHatEnterprise);
-    }
-
-    match name.as_str() {
-        "alpine linux" => Some(Type::Alpine),
-        "amazon linux" => Some(Type::Amazon),
-        "amazon linux ami" => Some(Type::Amazon),
-        "arch linux" => Some(Type::Arch),
-        "centos linux" => Some(Type::CentOS),
-        "centos stream" => Some(Type::CentOS),
-        "fedora" => Some(Type::Fedora),
-        "fedora linux" => Some(Type::Fedora),
-        "linux mint" => Some(Type::Mint),
-        "mariner" => Some(Type::Mariner),
-        "nixos" => Some(Type::NixOS),
-        "sles" => Some(Type::SUSE),
-        "ubuntu" => Some(Type::Ubuntu),
-        _ => None,
-    }
-}
-
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 struct ReleaseInfo<'a> {
-    os_type: Type,
+    // The release file the struct corresponds to
     path: &'a str,
-    version_matcher: Matcher,
+    // Outputs the os type given the release file
+    os_type: for<'b> fn(&'b str) -> Option<Type>,
+    // Outputs the os version given the release file
+    version: for<'b> fn(&'b str) -> Option<Version>,
```

Comment:
Could use more clarification for how function pointers work, especially with lifetimes.